### PR TITLE
Expand NT subclass to support SAM

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111078
* __->__ #109123

This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

File: torch/nested/_internal/ops.py [hunk 1]
```diff
@@ -172,19 +192,226 @@ def tensor_attr_unsupported_getter(func, *args, **kwargs):
...
+    )
+
+    inp = new_kwargs.pop("input")
+    weight = new_kwargs["weight"]
+    bias = new_kwargs["bias"]
+
+    new_values = torch.mm(inp._values, weight)
+    if bias is not None:
+        new_values += bias
+    return NestedTensor(new_values, inp._offsets)
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:206]: Unfortunately, for the symbolic singleton symint to be propagated, we should also need to thread the `inp._ragged_size` around. For now I'm using `extract_kwargs` as a utility to propagate all necessary metadata.

```suggestion
    return NestedTensor(new_values, **extract_kwargs(inp))
```

Comment (soulitzer) [reply torch/nested/_internal/ops.py:206]: Oh yeah this is probably rebase shenanigans, thanks for doing the rebase :)

File: torch/_prims_common/__init__.py [hunk 2]
```diff
@@ -1494,11 +1494,17 @@ def make_contiguous_strides_for(
     if not shape:
         return ()
 
+    def _ensure_at_least(a, thresh):
+        # sym_max() isn't supported for SingletonSymNode
+        if hasattr(a, "node") and a.node.singleton_int() is not None:
```
Comment (soulitzer) [inline torch/_prims_common/__init__.py:1499]: nit:
```suggestion
        if isinstance(a, torch.SymInt) and a.node.singleton_int() is not None:
```
hmmm interesting why its still going to the other path, you might want to check if threshold is also a singleton int?

**Review by soulitzer — APPROVED** [review]

File: torch/_prims_common/__init__.py (hunk 2 above)
Comment (jbschlosser) [reply torch/_prims_common/__init__.py]: Discussed offline: we need to check for SymInt wrapping a SingletonSymNode as well (this happens for the fake-ified case)

**Pull request closed without merging**